### ISSUE TITLE
Find the most comprehensive CC API

### ISSUE THREAD

**Issue by moonmath** [moonmath/static#13]
We need help finding the best public API. Created a gist that should help anyone with basic JavaScript skills look at the various public APIs out there.

https://gist.github.com/moonmath/30de25f4dd50b1d01e37e5c294097fb3

Please let me know if there's anything else I can look into further, or if there are any issues I'm ignoring related to this.

Binanace seems to be the most complete that I have found so far:

- total traded pairs: 250
- total USDT traded pairs: 6
- total base assets: 106

I'll use this information to implement as many CCs as possible at http://moonmath.win/ for as many CCs as possible.

**Comment by moonmath** [comment]
Found one that seems to be better:

https://www.alphavantage.co/documentation/

Looks like 393 base assets

**Comment by mr-highball** [comment]
I haven't seen alpha before, I'll take a look. If I were to implement this, the thought was to have data sources that are abstract enough to accommodate any API. Additionally, I would have a base data adapter which could allow for multiple export formats to be "plugged" into the collector (moonmath .csv, couchDB, mongo, sql, etc...). Having this pluggable infrastructure, the data collection service could be extended upon and used for a variety of purposes. Personally I like mongo over couch, but have used both with success.
Thoughts?

**Comment by moonmath** [comment]
I'm not concerned about consuming data from an API. The research required to identify ideal sources is time consuming, though.

I'm hoping to see more and better options than I have already found.

**Comment by mr-highball** [comment]
Perhaps an aggregation of API's would offer the best option? I know you wanted to try and avoid that, but it may lead to the most accurate picture. Binance + AlphaAdvantage + others.. 
For the API's that have overlapping pairs, the mean could be figured to provide a single closing price. If I get some time over the next couple of days I'll try and put something together.

**Comment by moonmath** [comment]
@mr-highball ,

That sounds about right. Realistically, we can't trust any one API. Especially true for alts.; price deviates too much between exchanges on some of them. If we're merging data between APIs I'll need to develop an algorithm that allows me to weight prices in some abstract way and publish those methods.

First thing is first: Evaluate APIs and get the list we're going to support. Then we'll be able to look at the data and figure out the definition for the merge algorithm.

**Comment by mr-highball** [comment]
@moonmath 
I added a comment to your gist about Bittrex. It looks to have about the same amount of pairs as Binance and should have a relatively low deviation for duplicate pairs.

**Comment by moonmath** [comment]
@mr-highball, The more I think about this the more I think it's a distinct project. The outages at binance made me realize that we do need something that's robust enough to tolerate API and exchange failure. We'll be using a separate language, and it's solving distinct and useful problems. I think we should start a new repo and do this work there.

You said your prefered language is Scala? Is that right?

**Comment by mr-highball** [comment]
I would agree on the separate project, and feel it could be used for various applications. My language of choice is Pascal. I started spec'ing out the design and can get a repo up pretty soon with ideas/issues. Sound ok?

**Comment by moonmath** [comment]
I went ahead an initialized a repo for you. I'm thinking about this as a distinct and configurable service. So, as generic as you can get. We've got a pretty good list of APIs to target initially, but you might want to throw in a handful of popular exchanges.

Generally, I'll follow your lead on this one because it seems like you have a vision of where you want it to go. It's critical that I'm able to pull output and stick it into the database in a consistent format. I'm thinking about this as a sort of Logstash for APIs: https://www.elastic.co/guide/en/logstash/current/introduction.html

If you have some better model for how this should work, please pass along a link.

**Comment by mr-highball** [comment]
Perfect. I like to use issues for the planning phase, so once I get my thoughts fully on paper, I'll transition those to github and loop you in on the discussion before working on any of the code.

**Comment by moonmath** [comment]
@mr-highball 

Hrm... It looks like logstash might have this problem licked already:

https://www.elastic.co/guide/en/logstash/current/plugins-inputs-http_poller.html

It looks like I can use this to poll any API and enforce a common format for their response.

Unless you can think of a reason this shouldn't work, we're back to assembling a list of APIs that we want to support.

**Comment by mr-highball** [comment]
I'm all for simple solutions. I personally don't have any experience with logstash, but looking over the doc, it appears like it would work.

**Comment by moonmath** [comment]
Yeah, I think this completes the picture:

https://www.elastic.co/guide/en/logstash/current/plugins-outputs-websocket.html

So, the CryptoCrunch project is better off as the logstash plugin configurations for various exchanges.

If you were jonesing to make this, don't let me hold you back, though.
